toml.md

```diff
@@ -417,8 +422,8 @@ str = ''''That,' she said, 'is still pointless.''''
 ```
 
 Control characters other than tab are not permitted in a literal string. Thus,
-for binary data, it is recommended that you use Base64 or another suitable ASCII
-or UTF-8 encoding. The handling of that encoding will be application-specific.
+for binary data, it is recommended that you use Base64 or another suitable
+binary-to-text encoding.
```

Comment:
@eksortso Please don't be so rash. If somebody makes a suggestion, wait for one or two days to give others a chance to comment before you actually act on it.

Regarding this sentence, i think it should stay. Since the spec mentions Base64, a quick reader might get the impression that a good TOML parser should support that encoding – this sentence makes it crystal clear that that's not the case.

Also, this change is entirely unrelated to Unicode. IF the sentence should be deleted (and, like I said, I don't think so) that should really be dealt with in a separate PR.

Reply:
@ChristianSi I am getting carried away, sorry. 24 hours after discussion for a change is a good rule of thumb here.

If I had my way, I would not mention binary data at all. Strings are not explicit binary sequences, and the people who are using strings to represent binary data don't rely on any unique features of TOML to use them. Let's just take all that out, and leave what remains: "Control characters other than tab are not permitted in a literal string."

Any new issue we encounter while addressing an old issue really should have its own PR, and it shouldn't drag the original PR into it. That is, unless we cannot separate out the changes. The original paragraph mentions "ASCII or UTF-8", so I'll argue that this change is tangential.

We could get rid of all references to ASCII in a separate PR, for instance. And I'm inclined to do so.

Reply:
The change from "ASCII or UTF-8 encoding" to "binary-to-text encoding" is fine (and was, at least in part, suggested by me). It's also clearly Unicode relevant since ASCII and UTF-8 were mentioned. What I object to is merely the deletion of the sentence "The handling of that encoding will be application-specific", which was a later and (IMHO) unwarranted change.

Reply:
That line was from the [original commit](https://github.com/toml-lang/toml/commit/2719d7f2822c63c9802afa32fa551b2927e05a7f) that added the "binary data" statement in the first place. The line about control codes not being allowed in literal strings was added later.

That particular commit does show some interesting history: the `\x` escape code, which was restored later, was taken out in the same commit. [At the time,](https://github.com/toml-lang/toml/issues/165#issuecomment-14471472) it was decided by @mojombo that strings would be authentic Unicode strings, and not byte sequences. When `\x` was brought back, it was within this context. But I can understand now why the language regarding binary data encoding was added.

I no longer want those lines purged. But perhaps, that whole sentiment ought to be reworded in more modern (and more correct) language.

Reply:
The most recent commit restates the binary data encoding language while discussing the escape codes. So this new language emphasizes that `\x`, `\u`, and `\U` are for Unicode, not for bytes. Mention of "application-specific" logic has been removed.

Reply:
Bear with me, @ChristianSi. I misread your objection to deleting the "application-specific" line. And @abelbraaksma you were suggesting we state that binary-to-text encodings are "implementation-specific," rather than application-specific.

Calling it implementation-specific is confusing, because as I have made clear, TOML has nothing to do with binary data. TOML parsers that implement Binary64 support are doing more than just parsing TOML. What consumers do with their strings is their business, not ours.

I'll add another line to my previous commit.

Reply:
What do you think of this replacement?

> Note that there is no low-level byte sequence syntax in TOML, and that the
> character escape codes must not be made to work in such a way. So for binary
> data, it is recommended that you use Base64 or another binary-to-text encoding,
> and that external tools ought to be used to interpret such encodings.

Reply:
That's pretty clear and complete, I think. I like it.